application/client/src/app/ui/tabs/sources/multiplefiles/component.ts

```diff
@@ -0,0 +1,201 @@
+                            this.filesUpdate.emit(this.state.files);
+                            this._countAndCheck();
+                        }
+                    })
+                    .catch((err: Error) => {
+                        this.log().error(`Fail to get file info: ${err.message}`);
+                    });
+            }
+        }
+    }
+
+    public ngOpenEach(files?: FileHolder[]) {
+        (files === undefined ? this.state.files : files).forEach((file: FileHolder) => {
+            if (file.selected) {
+                switch (file.type) {
+                    case FileType.Any:
+                    case FileType.Text:
+                        this.ilc()
+                            .services.system.opener.file(file.filename)
+                            .text()
+                            .catch((err: Error) => {
+                                this.log().error(`Fail to open text file; error: ${err.message}`);
+                            });
+                        break;
+                    case FileType.Dlt:
+                        this.ilc()
+                            .services.system.opener.file(file.filename)
+                            .dlt()
+                            .catch((err: Error) => {
+                                this.log().error(`Fail to open dlt file; error: ${err.message}`);
+                            });
+                        break;
+                }
+            }
+        });
+        this.tab.close();
+    }
+
+    public ngOnContext(event: IContextAction) {
+        switch (event.type) {
+            case EContextActionType.open:
+                if (event.files !== undefined) {
+                    this.ngOpenEach(event.files);
+                }
+                break;
+            case EContextActionType.update:
+                if (event.files !== undefined) {
+                    this._updateFiles(event.files);
+                }
+                this._countAndCheck();
+                break;
+        }
+    }
+
+    private _updateFiles(files: FileHolder[]) {
+        this.state.files =
+            files.length === 0
+                ? []
+                : this.state.files.filter((f: FileHolder) => !files.includes(f));
+    }
+
+    private _countAndCheck() {
+        this._selectedTypes = [];
+        let size = 0;
+        this.state.selectedCount = 0;
+        this.state.files.forEach((file: FileHolder) => {
+            if (file.selected) {
+                this.state.selectedCount++;
+                size += file.sizeInByte();
+                if (!this._selectedTypes.includes(file.type)) {
+                    this._selectedTypes.push(file.type);
+                }
+            }
+        });
+        this.state.selectedSize = bytesToStr(size);
+
+        this.state.openable = !this._selectedTypes.includes(FileType.Dlt);
+        this.state.concatable =
+            this._selectedTypes.length === 1 && this._selectedTypes.includes(FileType.Text);
+    }
```

Comment:
in this method, you almost manipulate with `state`. Would it be better to move this method just into state?